### DISCUSSION TITLE
Import Solution --skip-lower-version flag

### DISCUSSION THREAD

**Discussion by tonocode** [microsoft/powerplatform-build-tools#743] [Q&A] [answered]
In my production environment deployment pipeline, I haven't enabled the --skip-lower-version flag for the import solution task to allow for solution rollbacks. When I rerun the pipeline, the solution is successfully installed with a lower version number.

However, manually installing a solution with an older version throws an error "The import solution must have a higher version than the existing solution it is upgrading."

Apart from potentially going against best practices, could importing a solution with an older version cause any functional issues? Since --skip-lower-version is optional, I assumed allowing older versions to be installed wouldn't be an issue, but I'm concerned there might be functional reasons why Microsoft prevents this during manual installation.

**Comment by parvezghumra** [comment]
@tono0682 Admittedly, because the UI has never allowed an equivalent operation, I've (rightly or wrongly) assumed that it's just not possible to import a solution version that is lower than the version already installed. In addition, according to the ImportSolutionRequest object, there's no equivalent option available there: https://learn.microsoft.com/en-us/dotnet/api/microsoft.crm.sdk.messages.importsolutionrequest?view=dataverse-sdk-latest. So I thought this option was only for internal use and/or not working as expected yet.

But I realise it's available in the PAC CLI and exposed on the Import Solution PPBT Task and Power Platform Github Action. Interestingly, the default value of the parameter is false, which presumably means that if you don't specify a value for it in your workflow/pipeline (as I haven't been) and the solution you import is of a lower version than the one already installed in the target, then the lower version will be installed - effectively a 'rollback' of the solution (which I thought was never possible?!): 
https://learn.microsoft.com/en-us/power-platform/developer/cli/reference/solution#--skip-lower-version--slv
https://github.com/microsoft/powerplatform-actions/blob/main/import-solution/action.yml#L100 
https://github.com/microsoft/powerplatform-build-tools/blob/main/src/tasks/import-solution/import-solution-v2/task.json#L159

Am I understanding this correctly? 'Solution rollbacks' are possible, and in fact the default behaviour if you attempt to import previous versions of solutions, using PAC CLI, PPBT or Github Actions?

**Comment by parvezghumra** [comment]
Wow, just tested this using my Github Action workflows without explicitly setting the 'Skip Lower Version' parameter for 2 of my solutions, using the 2-stage (Import as Holding Solution + Apply Upgrade) upgrade process for each. And it actually rolls back the solution version!

I suspect PPBT in ADO will do the same. Anyone know the Power Platform Pipelines behaviour in this regard?

Another reason to automate deployments (if anyone needed more!) 😄

![image](https://github.com/microsoft/powerplatform-build-tools/assets/8039551/4da325fa-8cda-486b-90d3-1245a9be5796)

**Reply by tonocode** [reply]
Great to hear your pipelines yielded the same results as me! It's indeed really helpful for rollback purposes

**Comment by MattB-msft** [comment, marked as answer]
Sorry folks for the surprise here,  
This is a case where Doc's have not yet caught up with reality. 

### The change allows for an import of a lower version of a solution over a higher version,
__This is a modification to at the solution import behavior in general__.   
We have removed the check that blocked a lower version of a solution from being imported over an existing version via the API. 



When using "**Upgrade**" mode, either StageAndUpgrade, or Upgrade / Holding / Apply, just as it is with an upgrade it is a **Destructive Operation**.   

When using "**Update**" mode, it is **NOT** a Destructive Operation.  Its a merge operation as normal. 

**SDK/API's**: 
Pac CLI has the --Skip-Lower-Version (and you should put that in if you do not want the behavior of downgrade) 
Package Deployer still defaults to skipping lower versions, unless you override it, 
DevOps Tasks are based on PAC SOLUTION and PAC PACKAGE
**UX:** 
Studio UX currently blocks it but will be enabled in the future with guardrails.
Pipelines will soon support this with appropriate guardrails

**Reply by rajyraman** [reply]
@MattB-msft - Is the default value for "skip lower version" false? If so, does this mean that a solution can be imported as Upgrade even if the version is older, which was not the case before?

**Reply by SharonSmith-heuristicdev** [reply]
@MattB-msft - do you by any chance know when roughly this ability will be coming to PP Pipelines please?

**Reply by MattB-msft** [reply]
@rajyraman the default behavior will try to install a lower version,  SkipLowerVersion allows you to disabled that behavior. 
So yes, the default is 'false'

@SharonSmith-heuristicdev 
All I can say right now is **Soon**<sup>(tm) </sup>

**Reply by rajyraman** [reply]
@SharonSmith-heuristicdev - It is already available. https://github.com/microsoft/powerplatform-build-tools/blob/main/src/tasks/import-solution/import-solution-v2/task.json 
![image](https://github.com/microsoft/powerplatform-build-tools/assets/5035266/37ba6885-fbea-46e4-9d80-30b60ab51305)

**Reply by TomProkop** [reply]
I consider this a very bad decision. We used this platform behavior to prevent anyone accidentally uploading local or development builds to wrong environments.

**Reply by SharonSmith-heuristicdev** [reply]
Thank you for highlighting this @rajyraman.

**Reply by parvezghumra** [reply]
@SharonSmith-heuristicdev I think your question above addressed to @MattB-msft was asking about the ability to install previous versions of solution using in-platform pipelines which @MattB-msft said isn't available? @rajyraman's response above I think is highlighting the availability of this within Power Platform Build Tools for Azure DevOps if I'm not mistaken?

I haven't used in the in-platform pipelines so can't confirm whether this is already available there.

**Reply by parvezghumra** [reply]
@TomProkop Which bit are you referring to? Enabling this in the UI/Power Platform Pipelines or the the default behaviour (of 'Skip Lower Version' = false)?

**Reply by SharonSmith-heuristicdev** [reply]
Thanks for the clarification @parvezghumra.  Still useful knowledge, though, as we identify what we can and cannot yet do within the umbrella of the PP Pipelines, and what we'll need to turn to Azure DevOps for.  Thanks also for flagging Danny Styles' related LinkedIn post.  Had a useful exchange with Danny around his top tips in this area.

**Reply by rajyraman** [reply]
@parvezghumra @SharonSmith-heuristicdev - Hopefully Power Platform Pipelines is renamed in 2025 so that people like me don't confuse it with AzDO.

**Reply by caburk** [reply]
As Matt mentioned, it's a change in the solution import API to not block lower solution versions. The change was made to support a new "re-deploy" feature that's rolling out in pipelines in Power Platform. However, we wanted the same support for all solution imports, including the CLI, dev tools, manual import, etc. 

Due to the way our internal releases work, these don't all come online at the same time. The API and CLI came online first, and you'll see it as a native experience in PP pipelines shortly. Manual import will take a bit more time as it's pending some additional engineering work on our side (currently a client-side block is in place).

**Reply by parvezghumra** [reply]
That's really helpful insight. Thanks @caburk :)

**Reply by michaelmmcrae1** [reply]
Work-around (that has always worked) - import the Solution via the classic/legacy Solutions area :) 
Doesn't seem to matter Version higher or lower there
